subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/msvcpp/DefaultVisualStudioLocator.java

```diff
@@ -105,105 +100,149 @@ public SearchResult locateDefaultVisualStudioInstall(File candidate) {
+
+            File visualStudioDir = install.getInstallDir();
+            if (!foundInstalls.containsKey(visualStudioDir)) {
+                String displayVersion = install.getVersion() == VersionNumber.UNKNOWN ? "from system path" : install.getVersion().toString();
+                String displayCppVersion = install.getVisualCppVersion() == VersionNumber.UNKNOWN ? "from system path" : install.getVisualCppVersion().toString();
+                VisualCppInstall visualCpp = buildVisualCppInstall("Visual C++ " + displayCppVersion, visualStudioDir, visualCppDir, install.getVisualCppVersion(), install.getCompatibility());
+                VisualStudioInstall visualStudio = new VisualStudioInstall("Visual Studio " + displayVersion, visualStudioDir, install.getVersion(), visualCpp);
+                foundInstalls.put(visualStudioDir, visualStudio);
+            }
         }
-        LOGGER.debug("Found Visual C++ install {} using system path", visualCppDir);
-
-        File visualStudioDir = visualCppDir.getParentFile();
-        if (!foundInstalls.containsKey(visualStudioDir)) {
-            VisualCppInstall visualCpp = buildVisualCppInstall("Visual C++ from system path", visualStudioDir, visualCppDir, VersionNumber.UNKNOWN);
-            VisualStudioInstall visualStudio = new VisualStudioInstall(visualStudioDir, visualCpp);
-            foundInstalls.put(visualStudioDir, visualStudio);
-        }
-        pathInstall = foundInstalls.get(visualStudioDir);
     }
 
     private SearchResult locateUserSpecifiedInstall(File candidate) {
-        File visualStudioDir = FileUtils.canonicalize(candidate);
-        File visualCppDir = new File(visualStudioDir, "VC");
-        if (!isVisualStudio(visualStudioDir) || !isVisualCpp(visualCppDir)) {
-            LOGGER.debug("Ignoring candidate Visual C++ install for {} as it does not look like a Visual C++ installation.", candidate);
+        VisualStudioMetadata install = versionDeterminer.getVisualStudioMetadataFromInstallDir(candidate);
+
+        if (install != null) {
+            File visualStudioDir = install.getInstallDir();
+            File visualCppDir = install.getVisualCppDir();
+            if (!isValidInstall(install)) {
+                LOGGER.debug("Ignoring candidate Visual C++ install for {} as it does not look like a Visual C++ installation.", candidate);
+                return new InstallNotFound(String.format("The specified installation directory '%s' does not appear to contain a Visual Studio installation.", candidate));
+            }
+
+            if (!foundInstalls.containsKey(visualStudioDir)) {
+                String displayVersion = install.getVersion() == VersionNumber.UNKNOWN ? "from user provided path" : install.getVersion().toString();
+                String displayCppVersion = install.getVisualCppVersion() == VersionNumber.UNKNOWN ? "from user provided path" : install.getVisualCppVersion().toString();
```

Comment:
Same here, we should reject an install for which we can't figure out the version.

Reply:
Done.